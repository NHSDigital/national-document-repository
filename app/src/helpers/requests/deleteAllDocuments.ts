import axios from 'axios';
import { AuthHeaders } from '../../types/blocks/authHeaders';
import { DOCUMENT_TYPE } from '../../types/pages/UploadDocumentsPage/types';

type Args = {
    docType: DOCUMENT_TYPE;
    nhsNumber: string;
    baseUrl: string;
    baseHeaders: AuthHeaders;
};

type DeleteResponse = {
    data: string;
    status: number;
};
const deleteAllDocuments = async ({ docType, nhsNumber, baseUrl, baseHeaders }: Args) => {
    const gatewayUrl = baseUrl + '/DocumentDelete';

    try {
        const response: DeleteResponse = await axios.delete(gatewayUrl, {
            headers: {
                ...baseHeaders,
            },
            params: {
                patientId: nhsNumber,
                docType,
            },
        });
        return response;
    } catch (e) {
        throw e;
    }
};

export default deleteAllDocuments;
