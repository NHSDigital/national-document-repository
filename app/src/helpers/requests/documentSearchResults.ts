import { endpoints } from '../../types/generic/endpoints';
import { SearchResult } from '../../types/generic/searchResult';

import axios from 'axios';

type Args = {
    nhsNumber: string;
    baseUrl: string;
};

type GetDocumentSearchResultsResponse = {
    data: Array<SearchResult>;
};

const getDocumentSearchResults = async ({ nhsNumber, baseUrl }: Args) => {
    const gatewayUrl = baseUrl + endpoints.DOCUMENT_SEARCH;

    const { data }: GetDocumentSearchResultsResponse = await axios.get(gatewayUrl, {
        headers: {
            'Content-Type': 'application/json',
        },
        params: {
            patientId: nhsNumber,
        },
    });
    return data;
};

export default getDocumentSearchResults;
