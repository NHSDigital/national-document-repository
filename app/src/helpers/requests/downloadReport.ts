import { AuthHeaders } from '../../types/blocks/authHeaders';
import axios, { AxiosError } from 'axios';
import { ReportData } from '../../types/generic/reports';

type Args = {
    report: ReportData;
    fileType: string;
    baseUrl: string;
    baseHeaders: AuthHeaders;
};

type DownloadReportResponseData = {
    data: { url: string };
};

const downloadReport = async ({ report, fileType, baseUrl, baseHeaders }: Args): Promise<void> => {
    const gatewayUrl = baseUrl + report.endpoint;

    try {
        const { data }: DownloadReportResponseData = await axios.get(gatewayUrl, {
            headers: {
                ...baseHeaders,
            },
            params: {
                outputFileFormat: fileType,
            },
        });

        const link = document.createElement('a');
        link.href = data.url;
        link.setAttribute('download', '');

        link.click();
    } catch (e) {
        const error = e as AxiosError;
        throw error;
    }
};

export default downloadReport;
