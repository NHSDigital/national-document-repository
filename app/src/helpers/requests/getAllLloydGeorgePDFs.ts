import { AxiosError } from 'axios';
import axios from 'axios';
import { endpoints } from '../../types/generic/endpoints';
import { AuthHeaders } from '../../types/blocks/authHeaders';

type Args = {
    baseUrl: string;
    baseHeaders: AuthHeaders;
    nhsNumber: string;
};

const getAllLloydGeorgePDFs = async ({ baseUrl, baseHeaders, nhsNumber }: Args) => {
    const gatewayUrl = baseUrl + endpoints.LG_STITCH;
    try {
        const response = await axios.get(gatewayUrl, {
            headers: {
                ...baseHeaders,
            },
            params: {
                patientId: nhsNumber,
            },
        });
        return response;
    } catch (e) {
        const error = e as AxiosError;
        throw error;
    }
};

export default getAllLloydGeorgePDFs;
