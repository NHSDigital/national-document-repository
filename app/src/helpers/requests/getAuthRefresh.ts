import { AxiosInstance } from 'axios';
import { endpoints } from '../../types/generic/endpoints';
import { UserAuth } from '../../types/blocks/userAuth';

export type AuthTokenArgs = {
    refreshToken: string;
    axios: AxiosInstance;
};

const getAuthRefresh = async ({ refreshToken, axios }: AuthTokenArgs) => {
    try {
        const { data } = await axios.get(endpoints.REFRESH_AUTH, {
            params: { refreshToken },
        });
        const userAuth: UserAuth = data;
        return userAuth;
    } catch (e) {
        throw e;
    }
};

export default getAuthRefresh;
