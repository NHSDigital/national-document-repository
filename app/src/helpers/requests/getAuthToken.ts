import axios from 'axios';
import { endpoints } from '../../types/generic/endpoints';
import { UserAuth } from '../../types/blocks/userAuth';

export type AuthTokenArgs = {
    baseUrl: string;
    code: string;
    state: string;
};

const getAuthToken = async ({ baseUrl, code, state }: AuthTokenArgs) => {
    try {
        const tokenResponse: UserAuth = await axios.get(`${baseUrl}${endpoints.AUTH}`, {
            params: { code, state },
        });
        return tokenResponse;
    } catch (e) {
        throw e;
    }
};

export default getAuthToken;
