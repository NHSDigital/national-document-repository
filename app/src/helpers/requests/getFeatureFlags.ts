import { AuthHeaders } from '../../types/blocks/authHeaders';
import { endpoints } from '../../types/generic/endpoints';

import axios from 'axios';
import { FeatureFlags } from '../../types/generic/featureFlags';

type Args = {
    baseUrl: string;
    baseHeaders: AuthHeaders;
};

type GetFeatureFlagsResponse = {
    data: FeatureFlags;
};

export const defaultFeatureFlags = {
    testFeature1: false,
    testFeature2: false,
    testFeature3: false,
};

const getFeatureFlags = async ({ baseUrl, baseHeaders }: Args) => {
    const gatewayUrl = baseUrl + endpoints.FEATURE_FLAGS;
    try {
        const { data }: GetFeatureFlagsResponse = await axios.get(gatewayUrl, {
            headers: {
                ...baseHeaders,
            },
        });
        return data;
    } catch (e) {
        return defaultFeatureFlags;
    }
};

export default getFeatureFlags;
