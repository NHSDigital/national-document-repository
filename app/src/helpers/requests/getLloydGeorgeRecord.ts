import axios from 'axios';
import { AuthHeaders } from '../../types/blocks/authHeaders';
import { endpoints } from '../../types/generic/endpoints';
import waitForSeconds from '../utils/waitForSeconds';
import { JOB_STATUS } from '../../types/generic/downloadManifestJobStatus';
import { DownloadManifestError } from '../../types/generic/errors';
import { isRunningInCypress } from '../utils/isLocal';
export const DELAY_BETWEEN_POLLING_IN_SECONDS = isRunningInCypress() ? 0 : 10;

type Args = {
    nhsNumber: string;
    baseUrl: string;
    baseHeaders: AuthHeaders;
};

export type LloydGeorgeStitchResult = {
    jobStatus: string;
    numberOfFiles: number;
    totalFileSizeInByte: number;
    lastUpdated: string;
    presignedUrl: string;
};

type GetRequestArgs = {
    jobId: string;
    baseUrl: string;
    baseHeaders: AuthHeaders;
};

const ThreePendingErrorMessage = 'Failed to initiate download';
const UnexpectedResponseMessage =
    'Got unexpected response from server when trying to download record';

async function getLloydGeorgeRecord(args: Args): Promise<LloydGeorgeStitchResult> {
    const { nhsNumber, baseUrl, baseHeaders } = args;

    const jobId = await requestJobId(args);
    let pendingCount = 0;

    while (pendingCount < 3) {
        await waitForSeconds(DELAY_BETWEEN_POLLING_IN_SECONDS);
        const pollingResponse = await pollForPresignedUrl({
            baseUrl,
            baseHeaders,
            jobId,
        });

        switch (pollingResponse?.jobStatus) {
            case JOB_STATUS.COMPLETED:
                return pollingResponse;
            case JOB_STATUS.PROCESSING:
                continue;
            case JOB_STATUS.PENDING:
                pendingCount += 1;
                continue;
            default:
                throw new DownloadManifestError(UnexpectedResponseMessage);
        }
    }
    throw new DownloadManifestError(ThreePendingErrorMessage);
}

export const requestJobId = async ({ nhsNumber, baseUrl, baseHeaders }: Args): Promise<string> => {
    const gatewayUrl = baseUrl + endpoints.LLOYDGEORGE_STITCH;

    const response = await axios.post(gatewayUrl, '', {
        headers: {
            ...baseHeaders,
        },
        params: {
            patientId: nhsNumber,
        },
    });

    return response.data.jobId;
};
export const pollForPresignedUrl = async ({
    jobId,
    baseUrl,
    baseHeaders,
}: GetRequestArgs): Promise<LloydGeorgeStitchResult> => {
    const gatewayUrl = baseUrl + endpoints.LLOYDGEORGE_STITCH;

    const { data } = await axios.get(gatewayUrl, {
        headers: {
            ...baseHeaders,
        },
        params: {
            jobId,
        },
    });
    if (!data.presignedUrl.startsWith('https://')) {
        return Promise.reject({ response: { status: 500 } });
    }

    return {
        ...data,
        presignedUrl: `${data.presignedUrl}&origin=${
            typeof window !== 'undefined' ? window.location.href : ''
        }`,
    };
};
export default getLloydGeorgeRecord;
