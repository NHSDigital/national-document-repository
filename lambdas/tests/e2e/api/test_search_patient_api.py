import io
import logging
import os
import uuid

import requests
from tests.e2e.helpers.lloyd_george_data_helper import LloydGeorgeDataHelper

data_helper = LloydGeorgeDataHelper()

api_endpoint = os.environ.get("NDR_API_ENDPOINT")
api_key = os.environ.get("NDR_API_KEY")
dynamo_table = os.environ.get("NDR_DYNAMO_STORE") or ""


def test_search_patient_details(test_data, snapshot):
    lloyd_george_record = {}
    test_data.append(lloyd_george_record)

    lloyd_george_record["id"] = str(uuid.uuid4())
    lloyd_george_record["nhs_number"] = "9449305943"
    lloyd_george_record["data"] = io.BytesIO(b"Sample PDF Content")

    data_helper.create_metadata(lloyd_george_record)
    data_helper.create_resource(lloyd_george_record)

    url = f"https://{api_endpoint}/FhirDocumentReference?subject:identifier=https://fhir.nhs.uk/Id/nhs-number|{lloyd_george_record['nhs_number']}"
    headers = {
        "Authorization": "Bearer 123",
        "X-Api-Key": api_key,
        "X-Correlation-ID": "1234",
    }
    response = requests.request("GET", url, headers=headers)
    bundle = response.json()
    logging.info(bundle)

    del bundle["entry"][0]["resource"]["id"]
    del bundle["entry"][0]["resource"]["date"]
    del bundle["timestamp"]
    attachment_url = bundle["entry"][0]["resource"]["content"][0]["attachment"]["url"]
    assert (
        "https://internal-dev.api.service.nhs.uk/national-document-repository/DocumentReference/16521000000101~"
        in attachment_url
    )
    del bundle["entry"][0]["resource"]["content"][0]["attachment"]["url"]

    assert bundle["entry"][0] == snapshot


def test_multiple_cancelled_search_patient_details(test_data, snapshot):
    lloyd_george_record = {}
    test_data.append(lloyd_george_record)

    lloyd_george_record["id"] = str(uuid.uuid4())
    lloyd_george_record["nhs_number"] = "9449305943"
    lloyd_george_record["data"] = io.BytesIO(b"Sample PDF Content")
    lloyd_george_record["doc_status"] = "cancelled"

    data_helper.create_metadata(lloyd_george_record)
    data_helper.create_resource(lloyd_george_record)

    second_lloyd_george_record = {}
    test_data.append(second_lloyd_george_record)

    second_lloyd_george_record["id"] = str(uuid.uuid4())
    second_lloyd_george_record["nhs_number"] = "9449305943"
    second_lloyd_george_record["data"] = io.BytesIO(b"Sample PDF Content")
    second_lloyd_george_record["doc_status"] = "cancelled"

    data_helper.create_metadata(second_lloyd_george_record)
    data_helper.create_resource(second_lloyd_george_record)

    url = f"https://{api_endpoint}/FhirDocumentReference?subject:identifier=https://fhir.nhs.uk/Id/nhs-number|{lloyd_george_record['nhs_number']}"
    headers = {
        "Authorization": "Bearer 123",
        "X-Api-Key": api_key,
        "X-Correlation-ID": "1234",
    }
    response = requests.request("GET", url, headers=headers)
    bundle = response.json()

    del bundle["timestamp"]
    del bundle["entry"][0]["resource"]["id"]
    del bundle["entry"][0]["resource"]["date"]
    del bundle["entry"][0]["resource"]["content"][0]["attachment"]["url"]
    del bundle["entry"][1]["resource"]["id"]
    del bundle["entry"][1]["resource"]["date"]
    del bundle["entry"][1]["resource"]["content"][0]["attachment"]["url"]

    assert bundle["entry"][0] == snapshot
    assert bundle["entry"][1] == snapshot


def test_no_records(snapshot):
    lloyd_george_record = {}  # Initialize the dictionary
    lloyd_george_record["nhs_number"] = "9449305943"

    url = f"https://{api_endpoint}/FhirDocumentReference?subject:identifier=https://fhir.nhs.uk/Id/nhs-number|{lloyd_george_record['nhs_number']}"
    headers = {
        "Authorization": "Bearer 123",
        "X-Api-Key": api_key,
        "X-Correlation-ID": "1234",
    }
    response = requests.request("GET", url, headers=headers)
    bundle = response.json()

    assert bundle == snapshot


def test_invalid_patient(snapshot):
    lloyd_george_record = {}
    lloyd_george_record["nhs_number"] = "9999999993"

    url = f"https://{api_endpoint}/FhirDocumentReference?subject:identifier=https://fhir.nhs.uk/Id/nhs-number|{lloyd_george_record['nhs_number']}"
    headers = {
        "Authorization": "Bearer 123",
        "X-Api-Key": api_key,
        "X-Correlation-ID": "1234",
    }
    response = requests.request("GET", url, headers=headers)
    bundle = response.json()

    assert bundle == snapshot
