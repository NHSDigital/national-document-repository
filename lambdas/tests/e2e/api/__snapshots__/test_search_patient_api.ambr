# serializer version: 1
# name: test_invalid_patient
  dict({
    'issue': list([
      dict({
        'code': 'exception',
        'details': dict({
          'coding': list([
            dict({
              'code': 'invalid',
              'display': 'Invalid Content',
              'system': 'http://hl7.org/fhir/issue-type',
            }),
          ]),
        }),
        'diagnostics': 'Invalid patient number 9999999993',
        'severity': 'error',
      }),
    ]),
    'resourceType': 'OperationOutcome',
  })
# ---
# name: test_multiple_cancelled_search_patient_details
  dict({
    'resource': dict({
      'author': list([
        dict({
          'identifier': dict({
            'system': 'https://fhir.nhs.uk/Id/ods-organization-code',
            'value': 'H81109',
          }),
        }),
      ]),
      'content': list([
        dict({
          'attachment': dict({
            'contentType': 'application/pdf',
            'creation': '2023-01-01',
            'language': 'en-GB',
            'title': '1of1_Lloyd_George_Record_[Holly Lorna MAGAN]_[9449305943]_[29-05-2006].pdf',
          }),
        }),
      ]),
      'custodian': dict({
        'identifier': dict({
          'system': 'https://fhir.nhs.uk/Id/ods-organization-code',
          'value': 'H81109',
        }),
      }),
      'docStatus': 'cancelled',
      'resourceType': 'DocumentReference',
      'status': 'current',
      'subject': dict({
        'identifier': dict({
          'system': 'https://fhir.nhs.uk/Id/nhs-number',
          'value': '9449305943',
        }),
      }),
      'type': dict({
        'coding': list([
          dict({
            'code': '16521000000101',
            'display': 'Lloyd George record folder',
            'system': 'http://snomed.info/sct',
          }),
        ]),
      }),
    }),
  })
# ---
# name: test_multiple_cancelled_search_patient_details.1
  dict({
    'resource': dict({
      'author': list([
        dict({
          'identifier': dict({
            'system': 'https://fhir.nhs.uk/Id/ods-organization-code',
            'value': 'H81109',
          }),
        }),
      ]),
      'content': list([
        dict({
          'attachment': dict({
            'contentType': 'application/pdf',
            'creation': '2023-01-01',
            'language': 'en-GB',
            'title': '1of1_Lloyd_George_Record_[Holly Lorna MAGAN]_[9449305943]_[29-05-2006].pdf',
          }),
        }),
      ]),
      'custodian': dict({
        'identifier': dict({
          'system': 'https://fhir.nhs.uk/Id/ods-organization-code',
          'value': 'H81109',
        }),
      }),
      'docStatus': 'cancelled',
      'resourceType': 'DocumentReference',
      'status': 'current',
      'subject': dict({
        'identifier': dict({
          'system': 'https://fhir.nhs.uk/Id/nhs-number',
          'value': '9449305943',
        }),
      }),
      'type': dict({
        'coding': list([
          dict({
            'code': '16521000000101',
            'display': 'Lloyd George record folder',
            'system': 'http://snomed.info/sct',
          }),
        ]),
      }),
    }),
  })
# ---
# name: test_no_records
  dict({
    'issue': list([
      dict({
        'code': 'exception',
        'details': dict({
          'coding': list([
            dict({
              'code': 'not-found',
              'display': 'Not Found',
              'system': 'http://hl7.org/fhir/issue-type',
            }),
          ]),
        }),
        'diagnostics': 'Document reference not found',
        'severity': 'error',
      }),
    ]),
    'resourceType': 'OperationOutcome',
  })
# ---
# name: test_search_patient_details
  dict({
    'resource': dict({
      'author': list([
        dict({
          'identifier': dict({
            'system': 'https://fhir.nhs.uk/Id/ods-organization-code',
            'value': 'H81109',
          }),
        }),
      ]),
      'content': list([
        dict({
          'attachment': dict({
            'contentType': 'application/pdf',
            'creation': '2023-01-01',
            'language': 'en-GB',
            'title': '1of1_Lloyd_George_Record_[Holly Lorna MAGAN]_[9449305943]_[29-05-2006].pdf',
          }),
        }),
      ]),
      'custodian': dict({
        'identifier': dict({
          'system': 'https://fhir.nhs.uk/Id/ods-organization-code',
          'value': 'H81109',
        }),
      }),
      'docStatus': 'final',
      'resourceType': 'DocumentReference',
      'status': 'current',
      'subject': dict({
        'identifier': dict({
          'system': 'https://fhir.nhs.uk/Id/nhs-number',
          'value': '9449305943',
        }),
      }),
      'type': dict({
        'coding': list([
          dict({
            'code': '16521000000101',
            'display': 'Lloyd George record folder',
            'system': 'http://snomed.info/sct',
          }),
        ]),
      }),
    }),
  })
# ---
